File: backend/app/api/memory.py
"""Memory-card and scope-filtered RAG endpoints for authenticated users."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy import or_, select
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_authenticated_user
from app.errors import MEMORY_DISABLED, raise_api_error
from app.models.account import Account
from app.models.memory import Memory, MemoryCandidate
from app.models.user import User
from app.schemas.memory import (
    MemoryCandidateCreate,
    MemoryCandidateOut,
    MemoryConfirmRequest,
    MemoryOut,
    RAGSearchOut,
)
from app.services import memory as memory_service
from app.services import memory_sources, platform_features, rag
from app.services.space_fsm import is_active_member

router = APIRouter(tags=["memory-rag"])


def _candidate_out(
    db: Session, row: MemoryCandidate, identity: tuple[User, Account]
) -> MemoryCandidateOut:
    access = memory_sources.source_access(db, row, actor=identity[0], account=identity[1])
    return MemoryCandidateOut.model_validate(
        dict(
            id=row.id,
            source_message_id=row.source_message_id if access.readable else None,
            source_document_ref=row.source_document_ref if access.readable else None,
            source_span_json=memory_sources.public_source_snapshot(row) if access.readable else {},
            source_kind=row.source_kind,
            source_status=access.status,
            allowed_scopes=list(access.allowed_scopes),
            raw_quote=row.source_quote if access.readable else None,
            summary=row.summary if access.readable else None,
            purpose=row.purpose if access.readable else None,
            suggested_scope=row.suggested_scope,
            sensitivity=row.sensitivity,
            extractor_version=row.extractor_version,
            status=row.status,
            memory_id=row.memory_id,
            created_at=row.created_at,
            decided_at=row.decided_at,
        )
    )


def _memory_out(
    db: Session,
    row: Memory,
    identity: tuple[User, Account],
    space_id: int | None = None,
) -> MemoryOut:
    access = memory_sources.memory_access(
        db, row, actor=identity[0], account=identity[1], space_id=space_id
    )
    return MemoryOut.model_validate(
        dict(
            id=row.id,
            source_candidate_id=row.source_candidate_id,
            source_message_id=row.source_message_id if access.readable else None,
            source_document_ref=row.source_document_ref if access.readable else None,
            source_span_json=memory_sources.public_source_snapshot(row) if access.readable else {},
            source_kind=row.source_kind,
            source_status=access.status,
            allowed_scopes=list(access.allowed_scopes),
            raw_quote=row.raw_quote if access.readable else None,
            content=row.content if access.readable else None,
            purpose=row.purpose if access.readable else None,
            scope=row.scope,
            space_id=row.space_id,
            sensitivity=row.sensitivity,
            confirmation_status=row.confirmation_status,
            revision=row.revision,
            retention_until=row.retention_until,
            status=row.status,
            revoked_at=row.revoked_at,
            created_at=row.created_at,
            updated_at=row.updated_at,
        )
    )


@router.post("/memory-candidates", status_code=201, response_model=MemoryCandidateOut)
def create_memory_candidate(
    body: MemoryCandidateCreate,
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> MemoryCandidateOut:
    row = memory_service.create_candidate(
        db,
        account=identity[1],
        source_span=body.source_span,
        raw_quote=body.raw_quote,
        summary=body.summary,
        suggested_scope=body.suggested_scope,
        purpose=body.purpose,
        sensitivity=body.sensitivity,
        source_message_id=body.source_message_id,
        source_document_ref=body.source_document_ref,
        source=body.source.model_dump() if body.source is not None else None,
        idempotency_key=body.idempotency_key,
    )
    db.flush()
    result = _candidate_out(db, row, identity)
    result.model_dump_json()
    db.commit()
    return result


@router.get("/memory-candidates", response_model=list[MemoryCandidateOut])
def list_memory_candidates(
    include_decided: bool = Query(default=False),
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> list[MemoryCandidateOut]:
    if not platform_features.is_memory_enabled(db):
        raise_api_error(503, MEMORY_DISABLED, "Memory 功能未开启")
    stmt = select(MemoryCandidate).where(MemoryCandidate.author_account_id == identity[1].id)
    if not include_decided:
        stmt = stmt.where(MemoryCandidate.status == "pending")
    rows = db.scalars(stmt.order_by(MemoryCandidate.id.desc())).all()
    return [_candidate_out(db, row, identity) for row in rows]


@router.post("/memory-candidates/{candidate_id}/confirm", response_model=MemoryOut)
def confirm_memory_candidate(
    candidate_id: int,
    body: MemoryConfirmRequest,
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> MemoryOut:
    row = memory_service.confirm_candidate(
        db,
        candidate_id=candidate_id,
        confirmer=identity[0],
        confirmer_account=identity[1],
        scope=body.scope,
        retention_days=body.retention_days,
    )
    db.flush()
    result = _memory_out(db, row, identity)
    result.model_dump_json()
    db.commit()
    return result


@router.post("/memory-candidates/{candidate_id}/dismiss", response_model=MemoryCandidateOut)
def dismiss_memory_candidate(
    candidate_id: int,
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> MemoryCandidateOut:
    row = memory_service.dismiss_candidate(db, candidate_id=candidate_id, account_id=identity[1].id)
    db.flush()
    result = _candidate_out(db, row, identity)
    result.model_dump_json()
    db.commit()
    return result


@router.get("/memories", response_model=list[MemoryOut])
def list_memories(
    space_id: int | None = Query(default=None, gt=0),
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> list[MemoryOut]:
    if not platform_features.is_memory_enabled(db):
        raise_api_error(503, MEMORY_DISABLED, "Memory 功能未开启")
    actor, account = identity
    if space_id is None:
        memory_service.expire_due_memories(db, account_id=account.id)
    else:
        if not is_active_member(db, space_id, actor.id):
            return []
        memory_service.expire_due_memories(db, account_id=account.id, space_id=space_id)
    stmt = select(Memory).where(
        Memory.status == "active", Memory.confirmation_status == "confirmed"
    )
    private = (Memory.scope == "private") & (Memory.author_account_id == account.id)
    if space_id is None:
        stmt = stmt.where(private)
    else:
        shared = (Memory.space_id == space_id) & Memory.scope.in_(("household", "lineage"))
        stmt = stmt.where(or_(private, shared))
    rows = db.scalars(stmt.order_by(Memory.id.desc())).all()
    result = []
    for row in rows:
        projected = _memory_out(db, row, identity, space_id)
        # Only the owner can see quarantined metadata; other readers receive no row.
        if row.author_account_id == account.id or projected.source_status in (
            "available",
            "deleted_snapshot",
        ):
            projected.model_dump_json()
            result.append(projected)
    db.commit()
    return result


@router.post("/memories/{memory_id}/revoke", response_model=MemoryOut)
def revoke_memory(
    memory_id: int,
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> MemoryOut:
    row = memory_service.revoke_memory(db, memory_id=memory_id, account_id=identity[1].id)
    db.flush()
    result = _memory_out(db, row, identity)
    result.model_dump_json()
    db.commit()
    return result


@router.delete("/memories/{memory_id}", status_code=204)
def delete_memory(
    memory_id: int,
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> Response:
    memory_service.delete_memory(db, memory_id=memory_id, account_id=identity[1].id)
    db.commit()
    return Response(status_code=204)


@router.get("/rag/search", response_model=list[RAGSearchOut])
def search_rag(
    space_id: int = Query(gt=0),
    q: str = Query(min_length=1, max_length=500),
    limit: int = Query(default=10, ge=1, le=50),
    db: Session = Depends(get_db),
    identity: tuple[User, Account] = Depends(require_authenticated_user),
) -> list[RAGSearchOut]:
    rows = rag.search(db, actor=identity[0], space_id=space_id, query=q, limit=limit)
    result = [RAGSearchOut(**row.__dict__) for row in rows]
    for item in result:
        item.model_dump_json()
    db.commit()
    return result

File: frontend/src/api/memory.ts
import { apiClient } from '@/api/client'
import type {
  Memory,
  MemoryCandidate,
  MemoryCandidateSource,
  MemoryScope,
  MemoryScopeKind,
  MemorySensitivity,
  PlatformFeatureFlags,
  RagSearchResult,
} from '@/types/memory'

export async function fetchPlatformFeatures(): Promise<PlatformFeatureFlags> {
  const { data } = await apiClient.get<PlatformFeatureFlags>('/platform-features')
  return data
}

export interface ConfirmMemoryCandidatePayload {
  scope: MemoryScope
  retention_days?: number
}

/**
 * 手动新建候选（与后端 POST /memory-candidates 的既有合同对齐）：
 * 所有写入（含检索结果「保存」与私有记忆「新增」）都只能先创建候选，
 * 经待确认流程明确 scope 后才成为可检索记忆（V2.5 合同：无绕过审计的直接发布）。
 */
export interface CreateMemoryCandidatePayload {
  source: MemoryCandidateSource
  idempotency_key: string
  raw_quote?: string
  summary: string
  purpose: string
  suggested_scope: MemoryScopeKind
  sensitivity: MemorySensitivity
}

export async function fetchMemoryCandidates(includeDecided = false): Promise<MemoryCandidate[]> {
  const { data } = await apiClient.get<MemoryCandidate[]>('/memory-candidates', {
    params: includeDecided ? { include_decided: true } : undefined,
  })
  return data
}

export async function createMemoryCandidate(
  payload: CreateMemoryCandidatePayload,
): Promise<MemoryCandidate> {
  const { data } = await apiClient.post<MemoryCandidate>('/memory-candidates', payload)
  return data
}

/** The server returns private memories only without a space, and private plus shared memories with one. */
export async function fetchMemories(spaceId?: number): Promise<Memory[]> {
  const { data } = await apiClient.get<Memory[]>('/memories', {
    params: spaceId === undefined ? undefined : { space_id: spaceId },
  })
  return data
}

/** Confirmation is the only transition that makes a candidate searchable. */
export async function confirmMemoryCandidate(
  candidateId: number,
  payload: ConfirmMemoryCandidatePayload,
): Promise<Memory> {
  const { data } = await apiClient.post<Memory>(
    `/memory-candidates/${candidateId}/confirm`,
    payload,
  )
  return data
}

export async function dismissMemoryCandidate(candidateId: number): Promise<MemoryCandidate> {
  const { data } = await apiClient.post<MemoryCandidate>(
    `/memory-candidates/${candidateId}/dismiss`,
  )
  return data
}

export async function revokeMemory(memoryId: number): Promise<Memory> {
  const { data } = await apiClient.post<Memory>(`/memories/${memoryId}/revoke`)
  return data
}

export async function deleteMemory(memoryId: number): Promise<void> {
  await apiClient.delete(`/memories/${memoryId}`)
}

export async function searchRag(
  spaceId: number,
  query: string,
  limit = 20,
): Promise<RagSearchResult[]> {
  const { data } = await apiClient.get<RagSearchResult[]>('/rag/search', {
    params: { space_id: spaceId, q: query, limit },
  })
  return data
}

export const searchMemory = searchRag

const MEMORY_ERROR_COPY: Record<string, string> = {
  MEMORY_FEATURE_STATE_UNAVAILABLE: '能力状态暂时无法确认，请刷新后重试',
  MEMORY_DISABLED: '记忆功能当前未启用，请联系系统管理员',
  RAG_DISABLED: '检索与引用当前未启用，请联系系统管理员',
  MEMORY_REFRESH_FAILED: '操作已提交，但最新状态未能加载，请刷新或重试',
  MEMORY_CANDIDATE_NOT_FOUND: '候选记忆不存在或已被处理',
  MEMORY_SCOPE_FORBIDDEN: '来源或保存范围当前不可用，请确认读取权限并刷新后重试',
  MEMORY_SENSITIVE_SCOPE_FORBIDDEN: '敏感等级或共享范围不符合来源限制，请保留原敏感等级并缩小范围',
  MEMORY_STATE_CONFLICT: '记忆状态已变化，请刷新后重试',
  MEMORY_PRIVACY_BLOCKED: '隐私策略阻止了这个 scope，请选择更严格的范围',
  MEMORY_NOT_FOUND: '记忆不存在或已被删除',
  RAG_SOURCE_NOT_ALLOWED: '该材料未获准进入知识检索',
  RAG_POLICY_DENIED: '当前空间策略不允许检索这条知识',
  PROVIDER_LOCAL_REQUIRED_UNAVAILABLE: '该内容要求本地模型，但本地服务暂不可用',
}

export function friendlyMemoryError(code: string | null | undefined, fallback?: string): string {
  return code && code in MEMORY_ERROR_COPY
    ? MEMORY_ERROR_COPY[code] as string
    : fallback ?? '记忆操作失败，请稍后重试'
}
